In [3]:
import pandas as pd
import numpy as np
from sklearn.linear_model import Lasso
from sklearn.preprocessing import MinMaxScaler; scaler = MinMaxScaler()
from sklearn.metrics import mean_squared_error
from math import sqrt

On part d'un dataset contenant le taux de mortalité calculé par tranche d'âge et par sexe, pour chaque pays et pour chaque année. 
Les valeurs manquantes des indicateurs ont été remplacées par la dernière année disponible par pays.
data.csv est issu du programme xxx et il comprend :
    - une séléction d'années : 22 années (1995 - 2015)
    - tous les pays ayant une correpondance entre le fichier mortalité et le fichier worldbank : 94 pays
    - tous les types de cancer : 74 codes distincts
    - une séléction d'indicateurs

Liste des indicateurs, choisis pour avoir moins de 25% des valeurs manquantes :
        AG.LND.AGRI.K2	Agricultural land (sq. km)	0.096926511
        AG.LND.AGRI.ZS	Agricultural land (% of land area)	0.107572115
        AG.PRD.CROP.XD	Crop production index (2004-2006 = 100)	0.126373626
        AG.PRD.FOOD.XD	Food production index (2004-2006 = 100)	0.122510302
        AG.PRD.LVSK.XD	Livestock production index (2004-2006 = 100)	0.125429258
        AG.SRF.TOTL.K2	Surface area (sq. km)	0.01356456
        EN.ATM.CO2E.GF.ZS	CO2 emissions from gaseous fuel consumption (% of total)	0.148866758
        EN.ATM.CO2E.KT	CO2 emissions (kt)	0.158739698
        EN.ATM.CO2E.LF.KT	CO2 emissions from liquid fuel consumption (kt)	0.154447115
        EN.ATM.CO2E.LF.ZS	CO2 emissions from liquid fuel consumption (% of total)	0.148866758
        EN.ATM.CO2E.PC	CO2 emissions (metric tons per capita)	0.159168956
        EN.ATM.CO2E.SF.ZS	CO2 emissions from solid fuel consumption (% of total)	0.148866758
        EN.ATM.GHGO.KT.CE	Other greenhouse gas emissions, HFC, PFC and SF6 (thousand metric tons of CO2 equivalent)	0.107486264
        EN.ATM.GHGT.KT.CE	Total greenhouse gas emissions (kt of CO2 equivalent)	0.145861951
        EN.ATM.METH.AG.KT.CE	Agricultural methane emissions (thousand metric tons of CO2 equivalent)	0.154447115
        EN.ATM.METH.EG.KT.CE	Methane emissions in energy sector (thousand metric tons of CO2 equivalent)	0.154447115
        EN.ATM.METH.KT.CE	Methane emissions (kt of CO2 equivalent)	0.095896291
        EN.ATM.NOXE.AG.KT.CE	Agricultural nitrous oxide emissions (thousand metric tons of CO2 equivalent)	0.154447115
        EN.ATM.NOXE.EG.KT.CE	Nitrous oxide emissions in energy sector (thousand metric tons of CO2 equivalent)	0.154447115
        EN.ATM.NOXE.KT.CE	Nitrous oxide emissions (thousand metric tons of CO2 equivalent)	0.09220467
        EN.POP.DNST	Population density (people per sq. km of land area)	0.019059066
        IT.CEL.SETS	Mobile cellular subscriptions	0.123025412
        IT.MLT.MAIN	Fixed telephone subscriptions	0.144316621
        NY.ADJ.AEDU.GN.ZS	Adjusted savings: education expenditure (% of GNI)	0.054601648
        NY.GDP.FRST.RT.ZS	Forest rents (% of GDP)	0.152730082
        NY.GDP.MINR.RT.ZS	Mineral rents (% of GDP)	0.152730082
        NY.GDP.MKTP.CD	GDP (current US$)	0.165350275
        NY.GDP.PCAP.CD	GDP per capita (current US$)	0.16560783
        NY.GDP.TOTL.RT.ZS	Total natural resources rents (% of GDP)	0.154447115
        SE.PRE.DURS	Preprimary education, duration (years)	0.015024038
        SE.PRM.DURS	Primary education, duration (years)	0.010903159
        SE.SEC.DURS	Secondary education, duration (years)	0.025068681
        SP.ADO.TFRT	Adolescent fertility rate (births per 1,000 women ages 15-19)	0.066449176
        SP.DYN.AMRT.FE	Mortality rate, adult, female (per 1,000 female adults)	0.077781593
        SP.DYN.AMRT.MA	Mortality rate, adult, male (per 1,000 male adults)	0.077781593
        SP.DYN.CBRT.IN	Birth rate, crude (per 1,000 people)	0.051682692
        SP.DYN.CDRT.IN	Death rate, crude (per 1,000 people)	0.053571429
        SP.DYN.LE00.FE.IN	Life expectancy at birth, female (years)	0.066706731
        SP.DYN.LE00.IN	Life expectancy at birth, total (years)	0.066706731
        SP.DYN.TFRT.IN	Fertility rate, total (births per woman)	0.066277473
        SP.POP.DPND	Age dependency ratio (% of working-age population)	0.070484203
        SP.POP.GROW	Population growth (annual %)	0.005408654
        SP.URB.GROW	Urban population growth (annual %)	0.005236951
        SP.URB.TOTL.IN.ZS	Urban population (% of total)	0.004807692    

In [16]:
df3 = pd.read_csv('data.csv', header=0,sep=',', low_memory=False)

In [17]:
indicators_old = ['AG.LND.TOTL.K2',	'AG.LND.TOTL.UR.K2',	'AG.LND.AGRI.ZS',	'EG.ELC.ACCS.ZS',	'EN.ATM.CO2E.KT',	'NE.CON.TETC.CD',	'NE.EXP.GNFS.CD',	'NE.IMP.GNFS.CD',	'NV.AGR.TOTL.CD',	'NV.IND.MANF.CD',	'NV.SRV.TETC.CD',	'NY.GNS.ICTR.ZS',	'DT.ODA.OATL.CD',	'NE.CON.PETC.CD',	'NY.GDP.MKTP.CD',	'SP.POP.GROW',	'SP.POP.DPND',	'SP.POP.DPND.OL',	'SP.DYN.TFRT.IN',	'SP.DYN.TO65.FE.ZS',	'SP.DYN.TO65.MA.ZS',	'SP.DYN.IMRT.IN',	'SP.DYN.LE00.IN',	'SP.DYN.CONU.ZS',	'SN.ITK.DEFC.ZS',	'SH.XPD.TOTL.ZS',	'SH.TBS.INCD',	'SH.STA.MMRT.NE',	'SH.STA.BRTW.ZS',	'SH.PRV.SMOK.FE',	'SH.PRV.SMOK.MA',	'SH.MED.BEDS.ZS',	'SH.MED.NUMW.P3',	'SH.MED.PHYS.ZS',	'SH.HIV.1524.FE.ZS',	'SH.HIV.1524.MA.ZS',	'SM.POP.NETM',	'SL.UEM.TOTL.NE.ZS',	'SL.UEM.TOTL.ZS']
indicators = ['SP.URB.TOTL.IN.ZS',	'SP.URB.GROW',	'SP.POP.GROW',	'SE.PRM.DURS',	'AG.SRF.TOTL.K2',	'SE.PRE.DURS',	'EN.POP.DNST',	'SE.SEC.DURS',	'SP.DYN.CBRT.IN',	'SP.DYN.CDRT.IN',	'NY.ADJ.AEDU.GN.ZS',	'SP.DYN.TFRT.IN',	'SP.ADO.TFRT',	'SP.DYN.LE00.FE.IN',	'SP.DYN.LE00.IN',	'SP.POP.DPND',	'SP.DYN.AMRT.FE',	'SP.DYN.AMRT.MA',	'EN.ATM.NOXE.KT.CE',	'EN.ATM.METH.KT.CE',	'AG.LND.AGRI.K2',	'EN.ATM.GHGO.KT.CE',	'AG.LND.AGRI.ZS',	'AG.PRD.FOOD.XD',	'IT.CEL.SETS',	'AG.PRD.LVSK.XD',	'AG.PRD.CROP.XD',	'IT.MLT.MAIN',	'EN.ATM.GHGT.KT.CE',	'EN.ATM.CO2E.GF.ZS',	'EN.ATM.CO2E.LF.ZS',	'EN.ATM.CO2E.SF.ZS',	'NY.GDP.FRST.RT.ZS',	'NY.GDP.MINR.RT.ZS',	'EN.ATM.CO2E.LF.KT',	'EN.ATM.METH.AG.KT.CE',	'EN.ATM.METH.EG.KT.CE',	'EN.ATM.NOXE.AG.KT.CE',	'EN.ATM.NOXE.EG.KT.CE',	'NY.GDP.TOTL.RT.ZS',	'EN.ATM.CO2E.KT',	'EN.ATM.CO2E.PC',	'NY.GDP.MKTP.CD',	'NY.GDP.PCAP.CD']          

In [5]:
df3.columns
#dans le fichier data.csv utilisé ici, ce ne sont pas les indicateurs choisis en tenant compte des valeurs manquantes!!! 
#=> à refaire le fichier et à refaire tourner

Index(['Cancer_code', 'Sex', 'Age_bucket', 'Location', 'Year', 'Nb_deaths',
       'Country', 'Age_group', 'Population', 'Mortality_rate', 'area', 'year',
       'AG.LND.TOTL.K2', 'AG.LND.TOTL.UR.K2', 'AG.LND.AGRI.ZS',
       'EG.ELC.ACCS.ZS', 'EN.ATM.CO2E.KT', 'NE.CON.TETC.CD', 'NE.EXP.GNFS.CD',
       'NE.IMP.GNFS.CD', 'NV.AGR.TOTL.CD', 'NV.IND.MANF.CD', 'NV.SRV.TETC.CD',
       'NY.GNS.ICTR.ZS', 'DT.ODA.OATL.CD', 'NE.CON.PETC.CD', 'NY.GDP.MKTP.CD',
       'SP.POP.GROW', 'SP.POP.DPND', 'SP.POP.DPND.OL', 'SP.DYN.TFRT.IN',
       'SP.DYN.TO65.FE.ZS', 'SP.DYN.TO65.MA.ZS', 'SP.DYN.IMRT.IN',
       'SP.DYN.LE00.IN', 'SP.DYN.CONU.ZS', 'SN.ITK.DEFC.ZS', 'SH.XPD.TOTL.ZS',
       'SH.TBS.INCD', 'SH.STA.MMRT.NE', 'SH.STA.BRTW.ZS', 'SH.PRV.SMOK.FE',
       'SH.PRV.SMOK.MA', 'SH.MED.BEDS.ZS', 'SH.MED.NUMW.P3', 'SH.MED.PHYS.ZS',
       'SH.HIV.1524.FE.ZS', 'SH.HIV.1524.MA.ZS', 'SM.POP.NETM',
       'SL.UEM.TOTL.NE.ZS', 'SL.UEM.TOTL.ZS'],
      dtype='object')

In [10]:
#vérifier années présentes dans le dataset
sorted(df3['Year'].unique())
len(df3['Year'].unique())

22

In [9]:
#vérifier types de cancers présents dans le dataset
sorted(df3['Cancer_code'].unique())
len(df3['Cancer_code'].unique())

74

In [12]:
#vérifier les pays présents dans le dataset
sorted(df3['Country'].unique())
len(df3['Country'].unique())

94

In [15]:
#vérifier les indicateurs world bank présents dans le dataset
len(indicators) #39
len(df3.columns) #51

51

## Choix de type de cancer mieux expliqué à l'aide d'une régression lasso linéaire

In [78]:
nb_cancers = len(df3['Cancer_code'].unique()) #nombre de types de cancers dans le dataset
cancer_types = list(df3['Cancer_code'].unique()) #liste de types de cancers dans le dataset
#initialize lists
lasso_cancer_codes = np.empty(nb_cancers, dtype=object)
lasso_r2 = np.empty(nb_cancers, dtype='float64')
lasso_nb_features = np.empty(nb_cancers, dtype='int')
i=0 

In [79]:
for c in cancer_types:

    X_scaled = scaler.fit_transform( df3[df3['Cancer_code'] == c][indicators_old].dropna(axis=1, how='all').fillna(0) )
    y = df3[df3['Cancer_code'] == c]['Mortality_rate']
 
    linlasso = Lasso(alpha=0.005, max_iter = 10000).fit(X_scaled, y)  
    
    lasso_cancer_codes[i] = c
    lasso_nb_features[i] = np.sum(linlasso.coef_ != 0)
    lasso_r2[i] = linlasso.score(X_scaled, y)
    i = i+1

df_lasso1 = pd.DataFrame({'Cancer_code':lasso_cancer_codes, 'R2':lasso_r2, 'Nb_features':lasso_nb_features}) 

In [80]:
#df_lasso1.to_csv(path_or_buf = 'lasso_choice_cancers.csv', sep=',', na_rep='')
df_lasso1.sort_values(by=['R2'],inplace=False, ascending=False)

Cancer_code  Nb_features        R2
24         C17           33  0.068570
5      C00-C14           37  0.068442
18         C53           31  0.060277
26     C19-C21           33  0.046503
12         C43           20  0.042180
14         C71           27  0.041360
1          C56           32  0.039385
32         C68           11  0.034786
2          C64           27  0.034476
40     C33,C34           35  0.034372
65         D00           27  0.032666
22         C55           32  0.032358
10         C45           21  0.031941
8          C78           25  0.031088
46         C82           22  0.030634
61         D09            7  0.030008
16         C15           30  0.029643
58         C70           21  0.028950
69         D06            6  0.027907
42         C25           29  0.027085
28         C83           21  0.026749
13         C18           32  0.024540
6          C24           21  0.023931
47         C90           24  0.022840
44         C54           28  0.022838
41         C85           24  0.021835
43         C16           33  0.021323
59         C65           11  0.020115
0          C80           34  0.020004
54         C32           24  0.019790
..         ...          ...       ...
23         C76           25  0.013046
19     C40,C41           14  0.012700
52         C88           12  0.012665
67         D07           14  0.012643
56         C61           37  0.012401
55         C63           13  0.011663
49         C46            9  0.011603
29         C48           13  0.011581
20         C51           11  0.011499
25         C81           15  0.011466
66         D03            7  0.010718
45         C60           15  0.010175
34         C73           14  0.010159
48         C66            6  0.009228
73         C86            3  0.009141
71         D02           10  0.009016
3          C44           18  0.008994
64         C62            8  0.007397
9          C49           10  0.007180
36         C52            9  0.005437
17         C74            9  0.004967
57         D04            1  0.003817
60         C72            5  0.003482
27     C30,C31            8  0.003443
70         C58            2  0.002489
62         C69            5  0.002133
51         C75            5  0.001977
38         C39            6  0.001885
63         C37            3  0.001693
7          C47            2  0.001090

[74 rows x 3 columns]

Il paraît que le type de cancer le mieux expliqué par les variables choisies et pendant le période observée serait C17 (petit intestin), avec 33 features significatives et un R2 de 6,86%.
Ce résultat est très faible.

In [82]:
#Détails pour le meilleur modèle
c_series = df_lasso1[df_lasso1['R2'] == max(df_lasso1['R2'])]['Cancer_code'] #cancer code corresponding to best R2 score
c = list(c_series)[0]
X_scaled = scaler.fit_transform( df3[df3['Cancer_code'] == c][indicators_old].dropna(axis=1, how='all').fillna(0) )
y = df3[df3['Cancer_code'] == c]['Mortality_rate']
linlasso = Lasso(alpha=0.005, max_iter = 10000).fit(X_scaled, y)  
print('Type de cancer : ', c)
print('R2 : ', linlasso.score(X_scaled, y) )
print('Nombre de features prises en compte par le modèle : ', np.sum(linlasso.coef_ != 0))
print('Features non négatives ordonnées par coefficients :')
for e in sorted (list(zip(list(df3[indicators_old]), linlasso.coef_)),
                key = lambda e: -abs(e[1])):
    if e[1] != 0:
        print('\t{}, {:.3f}'.format(e[0], e[1]))   
        
y_pred = linlasso.predict(X_scaled)
rmse = sqrt(mean_squared_error(y, y_pred)); print('RMSE : ', rmse)
print('RMSE rapporté à la moyenne : ', rmse / np.mean(y))        

Type de cancer :  C17
R2 :  0.0685703764511
Nombre de features prises en compte par le modèle :  33
Features non négatives ordonnées par coefficients :
	SH.HIV.1524.FE.ZS, 52.744
	SP.DYN.TO65.FE.ZS, 44.976
	SH.HIV.1524.MA.ZS, -44.261
	NV.SRV.TETC.CD, 42.560
	NE.CON.PETC.CD, -31.964
	SP.DYN.TO65.MA.ZS, -28.696
	SN.ITK.DEFC.ZS, 21.704
	NV.AGR.TOTL.CD, -15.450
	SH.TBS.INCD, 15.048
	NV.IND.MANF.CD, -13.455
	EN.ATM.CO2E.KT, 12.815
	SL.UEM.TOTL.NE.ZS, -11.113
	SP.POP.DPND.OL, -10.734
	AG.LND.TOTL.K2, 8.789
	SP.DYN.IMRT.IN, -7.698
	NE.EXP.GNFS.CD, 7.672
	SP.POP.DPND, -5.128
	SH.XPD.TOTL.ZS, -4.831
	SP.POP.GROW, -4.456
	DT.ODA.OATL.CD, 4.264
	NY.GNS.ICTR.ZS, -2.922
	SH.MED.NUMW.P3, 2.644
	EG.ELC.ACCS.ZS, 2.462
	AG.LND.AGRI.ZS, -2.098
	SH.STA.MMRT.NE, -1.519
	SL.UEM.TOTL.ZS, -1.244
	SH.PRV.SMOK.MA, -1.214
	SP.DYN.CONU.ZS, -1.186
	SP.DYN.LE00.IN, -1.174
	SH.STA.BRTW.ZS, -0.686
	SH.MED.PHYS.ZS, 0.258
	SH.MED.BEDS.ZS, 0.135
	AG.LND.TOTL.UR.K2, 0.123
RMSE :  20.45419183376311
RMSE rapporté à la moy

## Choix de décalage optimal entre l'année d'observation des indicateurs socio-économiques et l'année d'observation du taux de mortalité (delta), à travers une régression lasso linéaire

On part d'un dataset contenant le taux de mortalité calculé par tranche d'âge et par sexe, pour chaque pays et pour chaque année. 
Les valeurs manquantes des indicateurs ont été remplacées par la dernière année disponible par pays.
data.csv est issu du programme xxx et il comprend :
    - une séléction d'années : 16 années (2000 - 2015)
    - tous les pays ayant une correpondance entre le fichier mortalité et le fichier worldbank : 99 pays (!!!pas le même nombre qu'avant!!!)
    - un type de cancer : C33, C34 (cancer du poumon); pas de variable type de cancer présente dans le dataset
    - une séléction d'indicateurs : 44 choisis parmis ceux avec moins de 25% de valeurs manquantes

In [52]:
df = pd.read_csv('data_2000.csv', header=0,sep=',', low_memory=False)
df.drop(['Unnamed: 0'], axis = 1, inplace = True)

In [62]:
indicators = ['SP.URB.TOTL.IN.ZS',	'SP.URB.GROW',	'SP.POP.GROW',	'SE.PRM.DURS',	'AG.SRF.TOTL.K2',	'SE.PRE.DURS',	'EN.POP.DNST',	'SE.SEC.DURS',	'SP.DYN.CBRT.IN',	'SP.DYN.CDRT.IN',	'NY.ADJ.AEDU.GN.ZS',	'SP.DYN.TFRT.IN',	'SP.ADO.TFRT',	'SP.DYN.LE00.FE.IN',	'SP.DYN.LE00.IN',	'SP.POP.DPND',	'SP.DYN.AMRT.FE',	'SP.DYN.AMRT.MA',	'EN.ATM.NOXE.KT.CE',	'EN.ATM.METH.KT.CE',	'AG.LND.AGRI.K2',	'EN.ATM.GHGO.KT.CE',	'AG.LND.AGRI.ZS',	'AG.PRD.FOOD.XD',	'IT.CEL.SETS',	'AG.PRD.LVSK.XD',	'AG.PRD.CROP.XD',	'IT.MLT.MAIN',	'EN.ATM.GHGT.KT.CE',	'EN.ATM.CO2E.GF.ZS',	'EN.ATM.CO2E.LF.ZS',	'EN.ATM.CO2E.SF.ZS',	'NY.GDP.FRST.RT.ZS',	'NY.GDP.MINR.RT.ZS',	'EN.ATM.CO2E.LF.KT',	'EN.ATM.METH.AG.KT.CE',	'EN.ATM.METH.EG.KT.CE',	'EN.ATM.NOXE.AG.KT.CE',	'EN.ATM.NOXE.EG.KT.CE',	'NY.GDP.TOTL.RT.ZS',	'EN.ATM.CO2E.KT',	'EN.ATM.CO2E.PC',	'NY.GDP.MKTP.CD',	'NY.GDP.PCAP.CD']          

In [57]:
#vérifier années présentes dans le dataset
sorted(df['Year'].unique())
print(len(df['Year'].unique()))

16


In [58]:
#vérifier les pays présents dans le dataset
sorted(df['Country'].unique())
len(df['Country'].unique())

99

In [64]:
#vérifier les indicateurs world bank présents dans le dataset
print(len(indicators)) #39
print(len(df.columns)) #51

44
49


In [66]:
print(sorted(indicators), '\n')
print(sorted(df.columns))

['AG.LND.AGRI.K2', 'AG.LND.AGRI.ZS', 'AG.PRD.CROP.XD', 'AG.PRD.FOOD.XD', 'AG.PRD.LVSK.XD', 'AG.SRF.TOTL.K2', 'EN.ATM.CO2E.GF.ZS', 'EN.ATM.CO2E.KT', 'EN.ATM.CO2E.LF.KT', 'EN.ATM.CO2E.LF.ZS', 'EN.ATM.CO2E.PC', 'EN.ATM.CO2E.SF.ZS', 'EN.ATM.GHGO.KT.CE', 'EN.ATM.GHGT.KT.CE', 'EN.ATM.METH.AG.KT.CE', 'EN.ATM.METH.EG.KT.CE', 'EN.ATM.METH.KT.CE', 'EN.ATM.NOXE.AG.KT.CE', 'EN.ATM.NOXE.EG.KT.CE', 'EN.ATM.NOXE.KT.CE', 'EN.POP.DNST', 'IT.CEL.SETS', 'IT.MLT.MAIN', 'NY.ADJ.AEDU.GN.ZS', 'NY.GDP.FRST.RT.ZS', 'NY.GDP.MINR.RT.ZS', 'NY.GDP.MKTP.CD', 'NY.GDP.PCAP.CD', 'NY.GDP.TOTL.RT.ZS', 'SE.PRE.DURS', 'SE.PRM.DURS', 'SE.SEC.DURS', 'SP.ADO.TFRT', 'SP.DYN.AMRT.FE', 'SP.DYN.AMRT.MA', 'SP.DYN.CBRT.IN', 'SP.DYN.CDRT.IN', 'SP.DYN.LE00.FE.IN', 'SP.DYN.LE00.IN', 'SP.DYN.TFRT.IN', 'SP.POP.DPND', 'SP.POP.GROW', 'SP.URB.GROW', 'SP.URB.TOTL.IN.ZS'] 

['AG.LND.AGRI.K2', 'AG.LND.AGRI.ZS', 'AG.PRD.CROP.XD', 'AG.PRD.FOOD.XD', 'AG.PRD.LVSK.XD', 'AG.SRF.TOTL.K2', 'Age_bucket', 'Country', 'EN.ATM.CO2E.GF.ZS', 'EN.ATM.CO2E.K

In [67]:
#initialize lists
max_delta = max(df['Year']) - min(df['Year'])
year = 2000
lasso_base_year = np.empty(max_delta, dtype='int') #year of features
lasso_delta = np.empty(max_delta, dtype='int') #year of target : year + delta
lasso_nb_obs = np.empty(max_delta, dtype='int') #number of observations the model is built on
lasso_r2 = np.empty(max_delta, dtype='float64')
lasso_nb_features = np.empty(max_delta, dtype='int')
i=0

In [68]:
for delta in range(1,max_delta+1):
    df['Mortality_shifted'] = df.groupby(['Sex', 'Age_bucket', 'Country'])['Mortality_rate'].shift(-delta)
    y = df[df['Year'] == 2000].dropna()['Mortality_shifted']
    X_scaled = scaler.fit_transform( df[df['Year'] == year].dropna()[indicators + ['Sex', 'Age_bucket']] )
    linlasso = Lasso(alpha=0.005, max_iter = 10000).fit(X_scaled, y)  

    lasso_base_year[i] = year
    lasso_delta[i] = delta    
    lasso_nb_obs[i] = len(y)               
    lasso_nb_features[i] = np.sum(linlasso.coef_ != 0)
    lasso_r2[i] = linlasso.score(X_scaled, y)
    i = i+1

df_lasso = pd.DataFrame({'Base_year':lasso_base_year,
                         'Delta':lasso_delta, 
                         'Nb_obs': lasso_nb_obs,
                         'Nb_features':lasso_nb_features,
                         'R2': lasso_r2 })   

In [69]:
df_lasso.sort_values(by=['R2'], ascending=False)

Base_year  Delta  Nb_features  Nb_obs        R2
14       2000     15           14     320  0.446156
5        2000      6           42    2016  0.358313
0        2000      1           41    2096  0.354860
13       2000     14           38    1312  0.354125
1        2000      2           43    2080  0.350789
4        2000      5           41    2016  0.348650
2        2000      3           40    2080  0.348255
3        2000      4           42    2048  0.347451
12       2000     13           40    1600  0.345936
9        2000     10           42    1888  0.344506
7        2000      8           41    2016  0.344372
6        2000      7           42    2016  0.343201
11       2000     12           39    1712  0.342550
8        2000      9           40    1984  0.342328
10       2000     11           41    1824  0.340498

Le meilleur résultat obtenu est un R2 de 0.446, pour delta = 15, c'est à dire les indicateurs socio-économiques de l'année 2000 expliquent le mieux le taux de mortalité de l'année 2015.

Cependant, le nombre d'observations pour ce décalage est très réduit(320), et seules 14 features sont retenues, ce qui pourrait indiquer un fort sur apprentissage.

De moins en moins d'observations sont disponibles pour les années les plus récentes pour le taux de mortalité, et il est nécessaire de les exclure du modèle.

Sinon, pour les autres décalages, le R2 oscille autour de 0.34

In [71]:
delta = df_lasso[df_lasso['R2'] == max(df_lasso['R2'])]['Delta']
delta

14    15
Name: Delta, dtype: int32

In [75]:
#Détails pour le meilleur modèle
delta_series = df_lasso[df_lasso['R2'] == max(df_lasso['R2'])]['Delta'] #delta corresponding to best R2 score
delta = list(delta_series)[0]

df['Mortality_shifted'] = df.groupby(['Sex', 'Age_bucket', 'Country'])['Mortality_rate'].shift(-delta)
y = df[df['Year'] == 2000].dropna()['Mortality_shifted']
X_scaled = scaler.fit_transform( df[df['Year'] == year].dropna()[indicators + ['Sex', 'Age_bucket']] )
linlasso = Lasso(alpha=0.005, max_iter = 10000).fit(X_scaled, y)   
print('Delta : ', delta)
print('R2 : ', linlasso.score(X_scaled, y) )
print('Nombre de features prises en compte par le modèle : ', np.sum(linlasso.coef_ != 0))
print('Features non négatives ordonnées par coefficients :')
for e in sorted (list(zip(list(df3[indicators_old]), linlasso.coef_)),
                key = lambda e: -abs(e[1])):
    if e[1] != 0:
        print('\t{}, {:.3f}'.format(e[0], e[1]))   
        
y_pred = linlasso.predict(X_scaled)
rmse = sqrt(mean_squared_error(y, y_pred)); print('RMSE : ', rmse)
print('RMSE rapporté à la moyenne : ', rmse / np.mean(y))

Delta :  15
R2 :  0.446155527658
Nombre de features prises en compte par le modèle :  14
Features non négatives ordonnées par coefficients :
	SH.XPD.TOTL.ZS, 31.448
	SH.MED.NUMW.P3, -30.594
	EN.ATM.CO2E.KT, -19.780
	SH.PRV.SMOK.FE, 13.046
	AG.LND.TOTL.UR.K2, -11.389
	AG.LND.AGRI.ZS, -10.199
	NY.GDP.MKTP.CD, -8.461
	NE.CON.PETC.CD, -7.950
	SH.TBS.INCD, -4.634
	SP.DYN.TO65.FE.ZS, -3.936
RMSE :  126.68531778331119
RMSE rapporté à la moyenne :  1.5665073862477357


On obtient un RMSE qui est supérieur à la moyenne du taux de mortalité, ce qui nous indique que le modèle obtient des résultats moins bons que un modèle dummy qui prédirait systématiquement la moyenne sans prendre en compte les features. 